precompiles/identity/src/lib.rs

```diff
@@ -0,0 +1,859 @@
+	<T as frame_system::Config>::AccountId,
+>>::Balance;
+
+/// Solidity selector of the Vote log, which is the Keccak of the Log signature.
+pub(crate) const SELECTOR_LOG_IDENTITY_SET: [u8; 32] = keccak256!("IdentitySet(address)");
+pub(crate) const SELECTOR_LOG_IDENTITY_CLEARED: [u8; 32] = keccak256!("IdentityCleared(address)");
+pub(crate) const SELECTOR_LOG_JUDGEMENT_REQUESTED: [u8; 32] =
+	keccak256!("JudgementRequested(address,uint32)");
+pub(crate) const SELECTOR_LOG_JUDGEMENT_UNREQUESTED: [u8; 32] =
+	keccak256!("JudgementUnrequested(address,uint32)");
+pub(crate) const SELECTOR_LOG_JUDGEMENT_GIVEN: [u8; 32] =
+	keccak256!("JudgementGiven(address,uint32)");
+pub(crate) const SELECTOR_LOG_SUB_IDENTITY_ADDED: [u8; 32] =
+	keccak256!("SubIdentityAdded(address,address)");
+pub(crate) const SELECTOR_LOG_SUB_IDENTITY_REMOVED: [u8; 32] =
+	keccak256!("SubIdentityRemoved(address,address)");
+pub(crate) const SELECTOR_LOG_SUB_IDENTITY_REVOKED: [u8; 32] =
+	keccak256!("SubIdentityRevoked(address)");
+
+/// A precompile to wrap the functionality from pallet-identity
+pub struct IdentityPrecompile<Runtime>(PhantomData<Runtime>);
+
+#[precompile_utils::precompile]
+#[precompile::test_concrete_types(mock::Runtime)]
+impl<Runtime> IdentityPrecompile<Runtime>
+where
+	Runtime: pallet_identity::Config + pallet_evm::Config,
+	Runtime::AccountId: Into<H160>,
+	Runtime::Hash: From<H256>,
+	Runtime::RuntimeCall: Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo,
+	<Runtime::RuntimeCall as Dispatchable>::RuntimeOrigin: From<Option<Runtime::AccountId>>,
+	Runtime::RuntimeCall: From<pallet_identity::Call<Runtime>>,
+	BalanceOf<Runtime>: TryFrom<U256> + Into<U256> + solidity::Codec,
+{
+	// Note: addRegistrar(address) & killIdentity(address) are not supported since they use a
+	// force origin.
+
+	// editorconfig-checker-disable
+	#[precompile::public("setIdentity((((bool,bytes),(bool,bytes))[],(bool,bytes),(bool,bytes),(bool,bytes),(bool,bytes),(bool,bytes),bool,bytes,(bool,bytes),(bool,bytes)))")]
+	fn set_identity(
+		handle: &mut impl PrecompileHandle,
+		info: IdentityInfo<Runtime::MaxAdditionalFields>,
+	) -> EvmResult {
+		// editorconfig-checker-enable
+		let caller = handle.context().caller;
+
+		let event = log1(
+			handle.context().address,
+			SELECTOR_LOG_IDENTITY_SET,
+			solidity::encode_event_data(Address(caller)),
+		);
+		handle.record_log_costs(&[&event])?;
+
+		let call = pallet_identity::Call::<Runtime>::set_identity {
+			info: Self::identity_to_input(info)?,
+		};
+
+		let origin = Runtime::AddressMapping::into_account_id(caller);
+		RuntimeHelper::<Runtime>::try_dispatch(handle, Some(origin).into(), call)?;
+
+		event.record(handle)?;
+
+		Ok(())
+	}
+
+	#[precompile::public("setSubs((address,(bool,bytes))[])")]
+	fn set_subs(
+		handle: &mut impl PrecompileHandle,
+		subs: BoundedVec<(Address, Data), Runtime::MaxSubAccounts>,
+	) -> EvmResult {
+		let subs: Vec<_> = subs.into();
```

Comment:
Could we just use `subs.0` from `BoundedVec` to avoid a copy?

Reply:
This isn't the substrate `BoundedVec` it's the precompile vec which only allows a `impl From` unfortunately

Reply:
The impl `From<BoundedVec<T, S>> for Vec<T>` already avoid copy